fcrepo-exts/fcrepo-migration-validator
# Update HeadOnly Validations

**JIRA Ticket**: https://fedora-repository.atlassian.net/browse/FCREPO-3833

# What does this Pull Request do?

Update the head only validations to be in line with head only migrations from migration-utils

# What's new?

* Created ValidationHandler interface for validating FedoraObjects 
* Split Head Only validations into new validation handler
* Updated head only validation for object existence/deletion
* Updated head only validation for object resource existence/deletion
* Update head only validation for number of object resource versions
* Removed head only logic from ValdatingObjectHandler

# How should this be tested?

Compile and verify all integration tests pass

Migrate a known good dataset (e.g. brown) with the head only flag enabled:
```
java -jar target/migration-utils-6.2.0-SNAPSHOT-driver.jar -d /data/migration/brown/repoarchive/datastreams_2019/ -o /data/migration/brown/repostore/data_2019/objects -a /data/migration/brown-headonly -i work/ -t legacy -H
```

Run the migration validator against the head only migration:
```
java -jar target/fcrepo-migration-validator-1.2.0-SNAPSHOT-driver.jar -s legacy -d /data/migration/brown/repoarchive/datastreams_2019 -o /data/migration/brown/repostore/data_2019/objects -c /data/migration/brown-headonly/data/ocfl-root -r /data/migration-results -H
```

Check the results of the validation report to make sure everything is working as expected.

# Additional Notes

When migrating head only, objects and datastreams which are marked as deleted will no longer show up in the F6 ocfl repository. The updates in the head only validations reflect this, though the wording for them might need to be improved.

I pulled some of the validations into the `ValidatingObjectHandler` which didn't need to be changed. The results of these are wrapped in Optionals even though some will always return a result -- I just wanted to keep everything consistent for them.

# Interested parties
@fcrepo/committers
